compatibility/velocity/src/main/java/com/ghostchu/compatibility/velocity/Main.java

```diff
@@ -0,0 +1,111 @@
+import com.velocitypowered.api.proxy.Player;
+import com.velocitypowered.api.proxy.ProxyServer;
+import com.velocitypowered.api.proxy.ServerConnection;
+import com.velocitypowered.api.proxy.messages.MinecraftChannelIdentifier;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.UUID;
+
+public final class Main {
+    private static final MinecraftChannelIdentifier QUICKSHOP_BUNGEE_CHANNEL = MinecraftChannelIdentifier.create("quickshop", "bungee");
+    private static final String SUB_CHANNEL_FORWARD = "forward";
+    private static final String SUB_CHANNEL_COMMAND = "command";
+    private static final String CHAT_COMMAND_REQUEST = "request";
+    private static final String CHAT_COMMAND_CANCEL = "cancel";
+    private final Set<UUID> pendingForward = Collections.synchronizedSet(new HashSet<>());
+
+    @Inject
+    private ProxyServer proxy;
+
+    @Subscribe
+    public void onProxyInitialization(ProxyInitializeEvent event) {
+        proxy.getChannelRegistrar().register(QUICKSHOP_BUNGEE_CHANNEL);
+    }
+
+    @Subscribe
+    public void onProxyShutdown(ProxyShutdownEvent event) {
+        this.pendingForward.clear();
+        proxy.getChannelRegistrar().unregister(QUICKSHOP_BUNGEE_CHANNEL);
+    }
+
+    @Subscribe
+    public void on(PluginMessageEvent event) {
+        if (!QUICKSHOP_BUNGEE_CHANNEL.equals(event.getIdentifier())) {
+            return;
+        }
+        ByteArrayDataInput in = event.dataAsDataStream();
+        String subChannel = in.readUTF();
+        if (SUB_CHANNEL_COMMAND.equalsIgnoreCase(subChannel)) {
+            // the receiver is a server when the proxy talks to a server
+            if (event.getSource() instanceof ServerConnection) {
+                String command = in.readUTF();
+                processCommand(command, in);
+            }
+        }
+    }
+
+    private void processCommand(String command, ByteArrayDataInput in) {
+        UUID uuid = UUID.fromString(in.readUTF());
+        switch (command) {
+            case CHAT_COMMAND_REQUEST -> this.pendingForward.add(uuid);
+            case CHAT_COMMAND_CANCEL -> this.pendingForward.remove(uuid);
+        }
+    }
+
+    private void forwardMessage(Player player, String message) {
+        ByteArrayDataOutput out = ByteStreams.newDataOutput();
+        out.writeUTF(SUB_CHANNEL_FORWARD);
+        out.writeUTF(message);
+        player.getCurrentServer().ifPresent(server ->
+            server.sendPluginMessage(QUICKSHOP_BUNGEE_CHANNEL, message.getBytes())
+        );
+    }
+
+    @Subscribe(order = PostOrder.FIRST)
+    public void onChat(PlayerChatEvent event) {
+        Player player = event.getPlayer();
+        // Workaround against kicking of players with valid signed key and client 1.19.1 or higher by velocity
+        if (player.getIdentifiedKey() != null
+                && player.getProtocolVersion().equals(ProtocolVersion.MINECRAFT_1_19_1)) {
+            return;
```

Comment:
If chat cannot be cancelled due the signing reason, we still can forward message instead just ignore the chat.